### PR TITLE
feat(packages): add Firebolt driver

### PR DESCRIPTION
**Pull request by Jeseeq**
**Check List**
- [x] Tests has been run in packages where changes made if available
- [x] Linter has been run for changed code
- [x] Tests for the changes have been added if not covered yet
- [x] Docs have been added / updated if required

**Description of Changes Made (if issue reference is not provided)**

### CODE REVIEW

File: docs/content/Configuration/Databases/Firebolt.mdx [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+### <--{"id" : "Setup"}--> Manual
+
+Add the following to a `.env` file in your Cube.js project:
+
+```bash
+CUBEJS_DB_NAME=firebolt_database
+CUBEJS_DB_USER=firebolt_user
+CUBEJS_DB_PASS=**********
+CUBEJS_FIREBOLT_ENGINE_ENDPOINT=engine_endpoint
```
Comment (cristipp) [inline docs/content/Configuration/Databases/Firebolt.mdx:20]: We have been using `CUBEJS_DB_URL` for other drivers (druid, elasticsearch) Can we use that for consistency?

File: packages/cubejs-testing/src/REQUIRED_ENV_VARS.ts [hunk 2]
```diff
@@ -15,6 +15,13 @@ export const REQUIRED_ENV_VARS: {[key: string]: string[]} = {
     'CUBEJS_DB_EXPORT_BUCKET',
     'CUBEJS_DB_BQ_CREDENTIALS',
   ],
+  firebolt: [
+    'CUBEJS_DB_USER',
+    'CUBEJS_DB_PASS',
+    'CUBEJS_DB_NAME',
+    'CUBEJS_FIREBOLT_ENGINE_ENDPOINT',
+    'CUBEJS_FIREBOLT_API_ENDPOINT'
```
Comment (cristipp) [inline packages/cubejs-testing/src/REQUIRED_ENV_VARS.ts:23]: I have not seen `CUBEJS_FIREBOLT_API_ENDPOINT` documented.

File: packages/cubejs-testing/test/driver-firebolt.test.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+import { executeTestSuiteFor } from './driver-test-suite';
+
+executeTestSuiteFor('firebolt');
```
Comment (cristipp) [inline packages/cubejs-testing/test/driver-firebolt.test.ts:3]: Sweet. Looks like this worked out of the box!

File: packages/cubejs-testing/package.json [hunk 4]
```diff
@@ -42,10 +42,13 @@
     "driver:athena:snap": "jest --verbose --updateSnapshot -i dist/test/driver-athena.test.js",
     "driver:bigquery": "jest --verbose -i dist/test/driver-bigquery.test.js",
     "driver:bigquery:snap": "jest --verbose --updateSnapshot -i dist/test/driver-bigquery.test.js",
+    "driver:firebolt": "jest --verbose -i dist/test/driver-firebolt.test.js",
```
Comment (cristipp) [inline packages/cubejs-testing/package.json:45]: Probably also want to add `driver:firebolt:snap` target here.

File: packages/cubejs-firebolt-driver/src/FireboltDriver.ts [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+    };
+
+    this.connectionParams = {
+      username: <string>process.env.CUBEJS_DB_USER,
+      password: <string>process.env.CUBEJS_DB_PASS,
+      database: <string>process.env.CUBEJS_DB_NAME,
+      engineEndpoint: <string>process.env.CUBEJS_FIREBOLT_ENGINE_ENDPOINT,
+    };
+
+    const apiEndpoint = process.env.CUBEJS_FIREBOLT_API_ENDPOINT;
```
Comment (cristipp) [inline packages/cubejs-firebolt-driver/src/FireboltDriver.ts:49]: What is the difference between `engineEndpoint` and `apiEndpoint`?

File: packages/cubejs-firebolt-driver/src/FireboltDriver.ts [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+    parameters: unknown[]
+  ): Promise<StreamTableData> {
+    const connection = await this.getConnection();
+
+    const statement = await connection.execute(query, {
+      settings: { output_format: OutputFormat.JSON },
+      parameters,
+    });
+
+    const { data: rowStream, meta: metaPromise } = await statement.streamResult();
```
Comment (cristipp) [inline packages/cubejs-firebolt-driver/src/FireboltDriver.ts:136]: Sweet, native streaming :)

File: packages/cubejs-firebolt-driver/src/FireboltDriver.ts [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+    const types = meta.map(({ type, name }) => ({ name, type: this.toGenericType(type) }));
+
+    return {
+      rowStream,
+      types,
+    };
+  }
+
+  public async unload(): Promise<DownloadTableCSVData> {
+    throw new Error('Unload is not supported');
```
Comment (cristipp) [inline packages/cubejs-firebolt-driver/src/FireboltDriver.ts:147]: Does Fireboolt support UNLOADing data to s3/gs?

**Review by cristipp — COMMENTED** [review]
Fantastic work! Thanks for the contribution!

File: docs/content/Configuration/Databases/Firebolt.mdx (hunk 1 above)
Comment (Jeseeq) [reply docs/content/Configuration/Databases/Firebolt.mdx:20]: snowflake or clickhouse use `CUBEJS_DB_NAME`
It is more like a name without port or host, not url

File: packages/cubejs-firebolt-driver/src/FireboltDriver.ts (hunk 5 above)
Comment (Jeseeq) [reply packages/cubejs-firebolt-driver/src/FireboltDriver.ts:49]: `apiEndpoint` is used for authentication (getting auth token). I decided to have it to be able to test on different environments of Firebolt (dev, staging). It is also not required

`engineEndpoint` is a http endpoint for querying database

File: packages/cubejs-firebolt-driver/src/FireboltDriver.ts (hunk 7 above)
Comment (Jeseeq) [reply packages/cubejs-firebolt-driver/src/FireboltDriver.ts:147]: I saw discussions about supporting unload few month ago. Not sure if it is ready. Need to research it

File: packages/cubejs-firebolt-driver/src/FireboltDriver.ts [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+  private getHydratedValue(value: unknown, meta: Meta) {
+    const { type } = meta;
+    if (isNumberType(type)) {
+      return `${value}`;
+    }
+    return value;
+  }
+
+  private hydrateRow = (row: Row, meta: Meta[]) => {
```
Comment (cristipp) [inline packages/cubejs-firebolt-driver/src/FireboltDriver.ts:128]: Out of curiosity, why did we add `FireboltDriver.hydrateRow`?

**Comment by cristipp** [comment]
Looks Good To Me.

**Pull request merged**